### ISSUE TITLE
openerp.fields.Integer() is currently limited to 32-bit integers

### ISSUE THREAD

**Issue by danielkauffman** [odoo/odoo#8437]
As of Odoo 8.0-20150824, openerp.fields.Integer() is limited to 32-bit integers.

Attempting to save integers larger than 32-bits raises DataError: integer out of range.

Some of the relevant code can be found in models.py around get_pg_type(), but the limit does not otherwise appear to be documented.

If one is not willing or able to patch Odoo core files, it is possible to work around this limitation by using openerp.fields.Text(), however, database operations on text fields are not as efficient as database operations on numeric fields.

It may be possible to increase the integer size simply by patching models.py, however, this change would affect every table and module in Odoo, and I have not tested this.  In particular, upgrading an existing installation may require locking the database and looping through each table to upgrade all of the integer fields.

Another option would be to patch fields.py to add a new class for large integers alongside the existing class for 32-bit integers.  However, supporting two separate integer data types in Odoo would add complexity to the API and might also be a premature optimization on modern systems.  (The differences between 32-bit and 64-bit integers, in terms of real-world performance and data storage requirements, are shrinking.)

There are many use cases for integers larger than 32-bits, for example, as Odoo interfaces with increasing numbers of third-party systems, any one of these systems may already be using integers larger than 32-bits, and the only viable option in Odoo currently is to use Text() fields to store these large integers.

Are there plans to increase the integer size in Odoo?

**Comment by rim-odoo** [comment]
Do you have any real-life business case where you can't use your ERP without 64-bits integers?

**Comment by mart-e** [comment]
We could change to big int (changing [this line](https://github.com/odoo/odoo/blob/f7aec21d210f4cef8327f7e65f249cdfd10f3f62/openerp/models.py#L164) should be enough) but of course we need a good reason to do so while 2.147.483.647 is probably enough for most integer cases. I am curious to know the answer to Richard's question.
If the change is needed, I would be afraid of the impact on the size and performances of the database to change that everywhere. So shouldn't probably be the default behaviour (an attribute on a field?)

By the way, Text field is not the only solution, float fields can be stored as numeric and then have [large precision](http://www.postgresql.org/docs/9.1/static/datatype-numeric.html).

**Comment by xmo-odoo** [comment]
@mart-e also note that XML-RPC only supports signed 32b integers, and JSON doesn't support integers at all, only IEEE 754-2008 double-precision numbers whose precise integral range is limited to [-(2<sup>53</sup>)+1, (2<sup>53</sup>)-1]

**Comment by 65HQqp5QTq8QwA3f** [comment]
One real-life case is integration between Odoo and eBay.  Several of the eBay APIs specify 64-bit integers and use values larger than 32-bits.  In the particular cases I have been working with, it has been possible to represent the fields in Odoo using Text() or Char().  However, as Odoo integrates with more and more third-party systems, especially in the enterprise market, more integrations will require the use of large integers (whether or not the third-party architecture is correct to use large integers), and Odoo currently has no way to represent such values natively.  (A float is of course fine in cases where numeric values are simply very large, but not great for an ID or any other exact value with infinite precision.  Even in cases where using a float could be made to work, it would not communicate the correct semantics, and would need to be cast to an integer every time it needed to be represented as an integer, which runs counter to the point of having Odoo handle the data model.)

Even if large integers where switched on globally, the performance and storage impact on modern machines is very likely to be minimal.  A database with several millions of records might increase in size by a few hundred megabytes, but even a $5 VPS comes with 20GB storage.  And most modern servers are 64-bit; IIRC most of the performance hit was back when 64-bit indexes where served from 32-bit machines, and even then the difference was minimal.

The only issue I see with switching globally would be XML-RPC, and probably the best way to deal with that is to translate large integers (or all integers) to text when exporting to XML-RPC, since that is unfortunately a limitation of the specification. But Odoo should not be limited due to XML-RPC, if an implementation wants to limit, that should be up the implementer, but out of the box I feel Odoo should support large integers.

This could even be a configuration file option, default to small integers, roll out as a beta feature, allow users to report bugs when the option is enabled, then at some point in the future, perhaps in several years as databases continue to grow, switch the default to large integers, but still support small integers for those who need them.

**Comment by mart-e** [comment]
Indeed, we had the same issue with our ebay connector but as it's an external identifier, no issue to use a char field. Anyway it's true that having an ID field reaching 2.1 billion records is not unachievable so I guess we will need to sort it out one day.
But of course we need to sort it out with XML/JSON RPC first.

**Comment by davidtranhp** [comment]
The real-life business case is when Odoo comes to countries with low rate currencies, 64bit Int should be required.
My country Vietnam is for an example. EUR/VND rate is about 24,405 today. If means we cannot set invoiced_target on a crm.team greater than 2,147,483,647 (~ only 87,993 EUR).

**Comment by Yenthe666** [comment]
Hi @mart-e,

I've hit this one too and I found another use case. Our customer generates very long integers for their internal numbering of products (for example 30011001804).
Perhaps it would be nice to add an attribute which allows us to save some fields as `big int`, while preserving the default 32 bit behaviour for the majority of fields.
@mart-e if you help me ahead I'll work out a PoC / PR.

Regards,
Yenthe

**Comment by mart-e** [comment]
@Yenthe666 I do not want to sound pessimistic but changing the field type is not trivial (actually should probably be a `fields.BigInt()` or something) and would probably not be accepted anyway.

`invoice_target` is kind of a weird case as it contains monetary amount in an integer field. Monetary amount should always be in float fields (where the issue does not appears), so this one can easily be avoided by changing the field type.

In the case of the product numbering (like barcode), we can probably avoid it using char field.

**Comment by Yenthe666** [comment]
Hi @mart-e,

Thanks for the quick feedback. 
Alright I can understand your point about the type not being trivial. Why would a new field not be accepted though? Couldn't we add this in Odoo V11?
The annoying part is that if you have a long integer and want to add it on a char field that the searching on it behaves different as for integer. So in my case I'd really love to use `int` instead of `char`.

Regards,
Yenthe

**Comment by mart-e** [comment]
We do not accept new field type if there isn't a real need. Having a new field type is something that needs to be maintain, adapt lot's of cases in the [orm](https://github.com/odoo/odoo/blob/0f05432374228df08a145bcd1c11ba3ef00301e5/odoo/addons/base/res/ir_property.py#L36), in the [addons](https://github.com/odoo/odoo/blob/0f05432374228df08a145bcd1c11ba3ef00301e5/addons/mail/models/mail_tracking_value.py#L39) and in the [web client](https://github.com/odoo/odoo/blob/0f05432374228df08a145bcd1c11ba3ef00301e5/addons/web/static/src/js/views/list_view.js#L1731) (and probably many modules). I am sure adding the monetary fields in v10 broke a lot of cases too.
This is lot of work and we won't do it only for a couple of use cases, sorry.

I do not say we will never reach the point where we need it but so far, we can avoid it.

**Comment by bluestingray** [comment]
@Yenthe666 is there any chance of coming back to this? Of either adding a `fields.BigInteger` or modifying `class Integer(Field):` to use `bigint`?

It may not be trivial, but I've tested ways of doing both and there's not a huge amount of modifications that need to happen to add support.

**Comment by mart-e** [comment]
@bluestingray what is the usecase ? As we said above, we do not consider the need at the moment as there is workaround (e.g. usually a float field).

**Comment by katyukha** [comment]
One more usecase for BigInt is to store filesizes of large files.

**Comment by pedrobaeza** [comment]
Why do you need to store it as integer? Store it as float or as string.

**Comment by katyukha** [comment]
Both, float field and char field, are workarounds.
- Char fields could be used only for values that are identifiers and no mathematical operations are planned on that values.
- Float fields could be used in most cases, but in this case it is required to keep in mind that it is float field.

For those cases when these workarounds are not suitable, I have created the [base_field_big_int](https://github.com/crnd-inc/generic-addons/tree/11.0/base_field_big_int) module with basic implementation of *BigInt* field for Odoo 11.0. May be it would be useful for somebody.

I got following results testing this addon:
- In general, it is working.
- JSON-RPC seems to be working (at least no errors raised and results contain correct values).
- Attempt to read *BigInt* field via XML-RPC raises error `int exceeds XML-RPC limits`
